algorithms/branching.py

```diff
@@ -0,0 +1,93 @@
+    isomorphism
+    :return: The number of isomorphisms if 'count_flag' is True or whether or not the graph has at least one
+    isomorphism if 'count_flag' is False
+    """
+    # Do color refinement on the graph
+    color_refinement(G)
+
+    is_balanced, is_bijected = is_balanced_or_bijected(G)
+    if not is_balanced:
+        # If the graph is unbalanced, the graph has no isomorphism
+        if not count_flag:
+            return False
+        return 0
+    if is_bijected:
+        # If the graph is balanced and bijected, the graph has exactly one isomorphism
+        if not count_flag:
+            return True
+        return 1
+
+    G_copy = G.copy()
+    colors, max_colornum = get_colors(G_copy)
+
+    # Choose the color C with at least 4 vertices of that color in the graph
+    for key in colors:
+        if len(colors[key]) >= 4:
+            C = key
+            continue
+
+    # Choose the first occurring vertex with color C in the list of vertices of the first graph
+    g0 = []
+    for v in colors[C]:
+        if v.graph_label == 1:
+            g0.append(v)
+    x = g0[0]
+    # Change the color of this vertex to a new color and append it to the list of fixed vertices for the first graph
+    x.colornum = max_colornum + 1
+    D.append(x)
+    # Create branches for all the possible fixed pairs of vertices for the chosen color
+    return __branching(G_copy, colors, C, D.copy(), I.copy(), count_flag)
+
+
+def __branching(G: 'Graph', colors: 'Dict[Int, List[Vertex]]', C: 'Int', D: 'List[Vertex]', I: 'List[Vertex]', count_flag: 'Bool'):
+    """
+    Creates branches of the graph (disjoint union of two graphs) and count the amount of isomorphisms for those graphs.
+    In one graph, one vertex of the color group is fixed. For each of the vertices in the other graph, a branch is
+    created fixing that vertex.
+    :param G: The graph (disjoint union of two graphs) to branch
+    :param colors: The colors present in the graph (disjoint union of two graphs) and its corresponding list of vertices
+    :param C: The chosen color group to create branches for
+    :param D: The list of vertices in one of the graphs that is fixed
+    :param I: The list of vertices in the other graph that is fixed
+    :param count_flag: Whether or not the amount of isomorphisms should be return or whether or not the graph has an
+    isomorphism
+    :return: The number of isomorphisms if 'count_flag' is True or whether or not the graph has at least one
+    isomorphism if 'count_flag' is False
+    """
+    # Create the list of vertices in the other graph with color C
+    g1 = []
+    for v in colors[C]:
+        if v.graph_label == 2:
+            g1.append(v)
+
+    num_isomorphisms = 0
+
+    # For each of the vertices in the list of vertices with color C, fix the vertex and change its color to the new
+    # color and determine the amount of isomorphisms for the resulting graph
+    for y0 in g1:
+        G_copy = G.copy()
```

Comment:
Ik weet dat het lijkt te werken, maar als ik hier logisch over nadenk, is iedere loop op regel 86 een nieuwe branch toch, en zou dus voor iedere nieuwe branch een nieuwe copy gemaakt moeten worden? Maar we gebruiken dezelfde kopie voor de gehele loop van regel 86. Is dat wel goed?

Reply:
We maken inderdaad wel 1 kopie die we aan alle volgende branches meegeven, maar we maken vrij snel al weer een kopie in de nieuwe ronde van count_isomorphisms. Het lijkt inderdaad te werken, maar zoals je het nu zegt zou de kleuring wel op diezelfde graaf moeten gebeuren en dat is niet wat je wilt. Misschien is het een idee om regel 34 naar bovenaan in de methode te plaatsen?

Reply:
Maar dan heb ik nog steeds het idee dat de kleuring die op regel 88 gebeurt, altijd op dezelfde G_copy gedaan wordt, dus dat in de eerste loop een vertex gekleurd wordt. En in de tweede loop nog 1, maar dat die eerste ook al veranderd is??

Reply:
> 
> 
> We maken inderdaad wel 1 kopie die we aan alle volgende branches meegeven, maar we maken vrij snel al weer een kopie in de nieuwe ronde van count_isomorphisms. Het lijkt inderdaad te werken, maar zoals je het nu zegt zou de kleuring wel op diezelfde graaf moeten gebeuren en dat is niet wat je wilt. Misschien is het een idee om regel 34 naar bovenaan in de methode te plaatsen?

Lijkt mij inderdaad beter om een kopie te hebben voordat er functies op worden uitgevoerd.
    color_refinement(G)
    is_balanced, is_bijected = is_balanced_or_bijected(G)

Worden nu uitgevoerd voor de .copy().

Reply:
Oh wacht, nu snap ik weer wat ik gedaan had. Vanaf regel 81 gaan we ervoor zorgen dat elke vertex van graaf 2 met kleur C een branch gaat vormen. Dus daar wordt voor elke y0 een kopie gemaakt, dat klopt. Daarna in de loop vanaf regel 86 gaat we op zoek naar de vertex in de kopie die hetzelfde label heeft als de vertex y0, zodat je voor elke branch de goede vertex kleurt. Dus uit die loop van regel 86 zal maar 1 branch komen, omdat we ervan uitgaat dat de labels van de vertices uniek zijn. Misschien als we een *continue* gebruiken in de loop vanaf 86, zodat we die niet vaker doorlopen dan nodig is (namelijk tot je de vertex gevonden hebt) of een andere structuur zodat wij zelf weten dat het geen volledige loop is.

Reply:
Je zou een teller kunnen bijhouden voor het aantal nog te vinden kleuren in graph 2.

Het aantal verwachtte kleuren is gelijk aan de lengte van de list g1 (r.72).  Of bedoel je iets anders?

Dit is meer een optimalisatie ding, beter om dit apart te testen lijkt mij.

Reply:
Ik snap wat je bedoelt, maar dat is denk ik ook meer een optimalisatie ding.